boa_engine/src/builtins/intl/date_time_format.rs

```diff
@@ -234,3 +259,1192 @@ pub(crate) fn to_date_time_options(
+                JsString::new("long"),
+            ]),
+        },
+        DateTimeComponents {
+            property: JsString::new("era"),
+            values: Vec::from([
+                JsString::new("narrow"),
+                JsString::new("short"),
+                JsString::new("long"),
+            ]),
+        },
+        DateTimeComponents {
+            property: JsString::new("year"),
+            values: Vec::from([JsString::new("2-digit"), JsString::new("numeric")]),
+        },
+        DateTimeComponents {
+            property: JsString::new("month"),
+            values: Vec::from([
+                JsString::new("2-digit"),
+                JsString::new("numeric"),
+                JsString::new("narrow"),
+                JsString::new("short"),
+                JsString::new("long"),
+            ]),
+        },
+        DateTimeComponents {
+            property: JsString::new("day"),
+            values: Vec::from([JsString::new("2-digit"), JsString::new("numeric")]),
+        },
+        DateTimeComponents {
+            property: JsString::new("dayPeriod"),
+            values: Vec::from([
+                JsString::new("narrow"),
+                JsString::new("short"),
+                JsString::new("long"),
+            ]),
+        },
+        DateTimeComponents {
+            property: JsString::new("hour"),
+            values: Vec::from([JsString::new("2-digit"), JsString::new("numeric")]),
+        },
+        DateTimeComponents {
+            property: JsString::new("minute"),
+            values: Vec::from([JsString::new("2-digit"), JsString::new("numeric")]),
+        },
+        DateTimeComponents {
+            property: JsString::new("second"),
+            values: Vec::from([JsString::new("2-digit"), JsString::new("numeric")]),
+        },
+        DateTimeComponents {
+            property: JsString::new("fractionalSecondDigits"),
+            values: Vec::from([
+                JsString::new("1.0"),
+                JsString::new("2.0"),
+                JsString::new("3.0"),
+            ]),
+        },
+        DateTimeComponents {
+            property: JsString::new("timeZoneName"),
+            values: Vec::from([
+                JsString::new("short"),
+                JsString::new("long"),
+                JsString::new("shortOffset"),
+                JsString::new("longOffset"),
+                JsString::new("shortGeneric"),
+                JsString::new("longGeneric"),
+            ]),
+        },
+    ])
+}
+
+/// Builds a list of `components::Bag` for all possible combinations of dateStyle and timeStyle
+/// ("full", "medium", "short", "long", undefined) for specified `locale` and `calendar`
+pub(crate) fn build_formats(locale: &JsString, calendar: &JsString) -> Vec<components::Bag> {
+    let locale_str = locale.to_string();
+    let locale = Locale::from_bytes(locale_str.as_bytes()).expect("Locale parsing failed");
+    let provider = InvariantDataProvider;
```

Comment:
Just a heads up. The doc for `InvariantDataProvider` specifies this provider is:
`Sometimes useful for testing. Not intended to be used in production environments.`

Currently we cannot pull the icu data until they release the 0.6.0 version of `icu_datagen` (which should be pretty soon, looking at https://github.com/unicode-org/icu4x/issues/1861), so I'd recommend to wait a bit to integrate this function until we have the required data to create our provider :)

Reply:
As far, as I can see from this PR https://github.com/boa-dev/boa/pull/2078 lookup_match tests must be updated. Should I file a new PR to do that?

Reply:
That should be done on that PR itself, just to have everything organized. We are still waiting on #2079 to fix the newest clippy lints. After that, I'll fix and rebase #2078 ASAP

Reply:
Just to clarify, I need to rebase to https://github.com/boa-dev/boa/pull/2083 once it is merged and then use `out_file` to get `DateTimeFormat`. Did I get it right?

Reply:
Almost. You just directly use the `ICU4X_DATA_PROVIDER` static.

Reply:
I don't think, I understand how to use `BoaProvider` properly. Replacing

```
let provider = InvariantDataProvider;
```
with
```
let provider = context.icu().provider();
```

results in 

```
the trait `icu_provider::ResourceProvider<WeekDataV1Marker>` is not implemented for `&dyn BoaProvider`
```

Reply:
Ah, I added to `BoaProvider` the minimal bounds to work. If there are some additional bounds required, you can add them to the definition of `BoaProvider`

Reply:
That bound is present in [this](https://github.com/boa-dev/boa/blob/518bad810974aa4cf6fcd3d864756f298d1b204d/boa_engine/src/context/icu.rs#L32) list. I assume, it has something to do with a reference to `dyn BoaProvider`. I've found several topics similar to [that](https://users.rust-lang.org/t/why-does-dyn-trait-not-implement-trait/30052) but none of them work for me.

Reply:
Ah, I see the problem now. You're passing `&provider` to the functions, but `provider` is a `&dyn BoaProvider`, meaning the function is testing trait bounds for `&&dynBoaProvider`. Remove the reference for `&provider` and it should work again

Reply:
That's the first thing I've tried:

```
721 |     provider,
    |     ^^^^^^^^ doesn't have a size known at compile-time

```

Reply:
That's probably a bug from the `icu` crate. I checked the function signature and the generic parameters implicitly require a `Sized` bound in them.

For now, can you manually implement the required bounds for `&dyn BoaProvider`?
I'll open an issue in the icu4x repo to address this, and we'll remove the implementations when the new version releases :)

Reply:
Sure. I'll give it a try.

Reply:
@jedel1043 I have tried different approaches, so that's my best shot so far. I have no idea where to get [source data](https://github.com/boa-dev/boa/pull/2072/files#diff-872037c107c01bf644ede412e4802b3eefeb5a70ce595c441f75651d45111b2aR44) though.

Is there a way to cast `&dyn BoaProvider` to a concrete type? I believe that's `icu_provider_blob::StaticDataProvider` in my case.

Reply:
No need, I fixed the impls on b2764e5 :)

Also, https://github.com/unicode-org/icu4x/pull/1969 has been merged, so we'll definitely remove those when 0.7.0 releases

Reply:
Great! Thanks a lot. I assume tests are failing on
```
let maybe_dtf = datetime::DateTimeFormat::<Gregorian>::try_new(locale, &provider, &options);
```

With

```
ICU4X data error: Missing data for variant or locale (key: datetime/lengths@1)
```

I'll try with another lengths shortly.

Reply:
Pro tip: enable the `log_error_context` feature on the `icu_provider` crate to see more detailed error messages in the log.